File: data/src/main/java/com/everis/data/repository/UserDataRepository.java
package com.everis.data.repository;

import android.support.annotation.Nullable;

import com.everis.data.model.FirebaseUserDataMapper;
import com.everis.data.network.firebase.FireBaseAuthImpl;
import com.everis.domain.model.User;
import com.everis.domain.repository.UserRepository;
import com.google.firebase.auth.FirebaseUser;

import rx.Observable;
import rx.functions.Func1;

/**
 * Created by everis on 25/04/18.
 */

public class UserDataRepository implements UserRepository {

    private FireBaseAuthImpl database;

    public UserDataRepository() {
        this.database = new FireBaseAuthImpl();
    }


    @Override
    public Observable<User> createUser(String email, String password) {
        return database.createUser(email, password).map(new Func1<FirebaseUser, User>() {
            @Override
            public User call(FirebaseUser firebaseUser) {
                return FirebaseUserDataMapper.transform(firebaseUser);
            }
        });
    }

    @Override
    public Observable<User> signInUser(String email, String password) {
        return database.signInUser(email, password).map(new Func1<FirebaseUser, User>() {
            @Override
            public User call(FirebaseUser firebaseUser) {
                return FirebaseUserDataMapper.transform(firebaseUser);
            }
        });
    }

    @Override
    public Observable<User> checkUserLoggedIn() {
        return database.checkUserLoggedIn().map(new Func1<FirebaseUser, User>() {
            @Override
            public User call(FirebaseUser firebaseUser) {
                return FirebaseUserDataMapper.transform(firebaseUser);
            }
        });
    }


    @Override
    public Observable<User> getUserState() {
        return database.observeAuthStateChange().map(new Func1<FirebaseUser, User>() {
            @Override
            public User call(@Nullable FirebaseUser firebaseUser) {
                return FirebaseUserDataMapper.transform(firebaseUser);
            }
        });
    }

    @Override
    public Observable<Void> logOut() {
        return database.logOut();
    }

    @Override
    public Observable<Void> deleteUser() {
        return database.deleteUser();
    }


}

File: data/src/main/java/com/everis/data/repository/UserDataRepository.kt
package com.everis.data.repository

import com.everis.data.model.FirebaseUserDataMapper
import com.everis.data.network.firebase.FireBaseAuth
import com.everis.data.network.firebase.FireBaseAuthImpl
import com.everis.data.network.firebase.FireBaseStorage
import com.everis.data.network.firebase.FireBaseStorageImpl
import com.everis.domain.model.User
import com.everis.domain.repository.UserRepository
import com.google.firebase.storage.FirebaseStorage

import rx.Observable
import java.io.InputStream

class UserDataRepository : UserRepository {


    private val database: FireBaseAuth = FireBaseAuthImpl()
    private val storage: FireBaseStorage = FireBaseStorageImpl()

    override fun createUser(email: String, password: String): Observable<User> {
        return database.createUser(email, password).map { firebaseUser -> FirebaseUserDataMapper.transform(firebaseUser) }
    }

    override fun signInUser(email: String, password: String): Observable<User> {
        return database.signInUser(email, password).map { firebaseUser -> FirebaseUserDataMapper.transform(firebaseUser) }
    }

    override fun checkUserLoggedIn(): Observable<User> {
        return database.checkUserLoggedIn().map { firebaseUser -> FirebaseUserDataMapper.transform(firebaseUser) }
    }

    override fun getUserState(): Observable<User> {
        return database.observeAuthStateChange().map { firebaseUser -> FirebaseUserDataMapper.transform(firebaseUser) }
    }

    override fun logOut(): Observable<Void> {
        return database.logOut()
    }

    override fun deleteUser(): Observable<Void> {
        return database.deleteUser()
    }

    override fun uploadPicture(id: String,stream : InputStream): Observable<String> {
        return storage.uploadFile(FirebaseStorage.getInstance().reference.child("pictures").child(id),stream)
    }



}
